### PR TITLE
Handle a failing payment case

### PR DESCRIPTION
**Pull request by jakubtobiasz**
part of #52

### CODE REVIEW

File: config/services/payment.php [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+declare(strict_types=1);
+
+namespace Symfony\Component\DependencyInjection\Loader\Configurator;
+
+use CommerceWeavers\SyliusTpayPlugin\Payment\Canceller\PaymentCanceller;
+use CommerceWeavers\SyliusTpayPlugin\Payment\Canceller\PaymentCancellerInterface;
+use CommerceWeavers\SyliusTpayPlugin\Payment\Checker\PaymentCancellationPossibilityChecker;
+use CommerceWeavers\SyliusTpayPlugin\Payment\Checker\PaymentCancellationPossibilityCheckerInterface;
+
+return function(ContainerConfigurator $container): void {
```
Comment (arti0090) [inline config/services/payment.php:12]: could also be `static`

**Review by arti0090 — APPROVED** [review]

File: config/services.php [hunk 2]
```diff
@@ -13,8 +13,6 @@
         ->set('env(TPAY_CLIENT_SECRET)', '')
         ->set('env(TPAY_CARDS_API)', '')
         ->set('env(TPAY_NOTIFICATION_SECURITY_CODE)', '')
-        ->set('env(TPAY_MERCHANT_ID)', '')
-        ->set('env(TPAY_GOOGLE_MERCHANT_ID)', '')
```
Comment (coldic3) [inline config/services.php:17]: Can we remove all parameters from here as they are defined in the `config/config.php`?

File: phpstan.neon [hunk 3]
```diff
@@ -22,3 +22,10 @@ parameters:
...
+        - '#Property CommerceWeavers\\SyliusTpayPlugin\\Payment\\Checker\\PaymentCancellationPossibilityChecker::\$stateMachine has unknown class Sylius\\Abstraction\\StateMachine\\StateMachineInterface as its type\.#'
+        - '#Call to method can\(\) on an unknown class Sylius\\Abstraction\\StateMachine\\StateMachineInterface\.#'
+        - '#Parameter \$stateMachine of method CommerceWeavers\\SyliusTpayPlugin\\Payment\\Canceller\\PaymentCanceller::__construct\(\) has invalid type Sylius\\Abstraction\\StateMachine\\StateMachineInterface\.#'
+        - '#Property CommerceWeavers\\SyliusTpayPlugin\\Payment\\Canceller\\PaymentCanceller::\$stateMachine has unknown class Sylius\\Abstraction\\StateMachine\\StateMachineInterface as its type\.#'
+        - '#Caught class Sylius\\Abstraction\\StateMachine\\Exception\\StateMachineExecutionException not found\.#'
+        - '#Call to method apply\(\) on an unknown class Sylius\\Abstraction\\StateMachine\\StateMachineInterface\.#'
```
Comment (coldic3) [inline phpstan.neon:31]: Why? Do you know what's the problem behind that?

File: src/Payment/Canceller/PaymentCanceller.php [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+use SM\Factory\FactoryInterface;
+use Sylius\Abstraction\StateMachine\Exception\StateMachineExecutionException;
+use Sylius\Abstraction\StateMachine\StateMachineInterface;
+use Sylius\Component\Core\Model\PaymentInterface;
+use Sylius\Component\Payment\PaymentTransitions;
+
+final class PaymentCanceller implements PaymentCancellerInterface
+{
+    public function __construct(
+        private readonly StateMachineInterface|null $stateMachine,
```
Comment (coldic3) [inline src/Payment/Canceller/PaymentCanceller.php:17]:
```suggestion
        private readonly ?StateMachineInterface $stateMachine,
```

just to follow the convention

File: src/Payment/Exception/PaymentCannotBeCancelledException.php [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace CommerceWeavers\SyliusTpayPlugin\Payment\Exception;
+
+use Sylius\Component\Core\Model\PaymentInterface;
+
+class PaymentCannotBeCancelledException extends \RuntimeException
+{
+    public function __construct(
+        PaymentInterface $payment,
+        string $message = 'Payment with id "%d" cannot be cancelled.',
```
Comment (coldic3) [inline src/Payment/Exception/PaymentCannotBeCancelledException.php:13]:
```suggestion
        string $message = 'Payment with ID "%d" cannot be cancelled.',
```

File: config/services.php (hunk 2 above)
Comment (jakubtobiasz) [reply config/services.php:17]: Rebase left-over 😅

File: phpstan.neon (hunk 3 above)
Comment (jakubtobiasz) [reply phpstan.neon:31]: It does not exist on Sylius `1.12`, so I silence it; it's strictly related to supporting two Sylius versions at the same time

File: tests/E2E/RetryPayment/RetryingPaymentTest.php [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+namespace E2E\RetryPayment;
+
+use Tests\CommerceWeavers\SyliusTpayPlugin\Api\Utils\OrderPlacerTrait;
+use Tests\CommerceWeavers\SyliusTpayPlugin\E2E\E2ETestCase;
+use Tests\CommerceWeavers\SyliusTpayPlugin\E2E\Helper\Order\RetryPaymentTrait;
+
+final class RetryingPaymentTest extends E2ETestCase
+{
+    use OrderPlacerTrait;
```
Comment (jakubtobiasz) [inline tests/E2E/RetryPayment/RetryingPaymentTest.php:13]: `OrderExampleFactory` has been making troubles; I've reused order placer from the API and it works xD

File: phpstan.neon (hunk 3 above)
Comment (coldic3) [reply phpstan.neon:31]: Ok, I see. However, we could take advantage of that it's a regex:
```suggestion
        - '#has invalid type Sylius\\Abstraction\\StateMachine\\StateMachineInterface#'
        - 'has unknown class Sylius\\Abstraction\\StateMachine\\StateMachineInterface#'
        - 'an unknown class Sylius\\Abstraction\\StateMachine\\StateMachineInterface#'
        - '#Caught class Sylius\\Abstraction\\StateMachine\\Exception\\StateMachineExecutionException not found\.#'
        - '#Call to method apply\(\) on an unknown class Sylius\\Abstraction\\StateMachine\\StateMachineInterface\.#'
```

**Review by coldic3 — APPROVED** [review]

**Pull request merged**